### PR TITLE
Fix bug of sending empty log from docker container

### PR DESCRIPTION
**Pull request by tt810**
### What does this PR do?

* Add condition of checking the docker header when form a log line in logs decoder.
* Rename DockerParser to Parser to pass linter check.

### Motivation

Client complain

### Additional Notes

Tested with client failing user case.

### CODE REVIEW

File: releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Sometime agent sends empty logs when running with docker container.
```
Comment (NBParis) [inline releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml]: Let's try to give more details if we can in release notes. this is what is being read by a lot of people that do not have much context about the situation.

Suggestion:

```
Empty logs could be appear when collecting Docker logs in addition of the actual container logs. This was due to the way the Agent handles the header Docker adds to the logs.
The process has been changed to make sure that no empty logs are generated.
```

WDYT?

File: releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+fixes:
+  - |
+    Empty logs could be appear when collecting Docker logs in addition
```
Comment (achntrl) [inline releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml]: I would give that much technical details and just say `Fix an issue where empty logs could be sent when collection Docker logs`

**Review by achntrl — COMMENTED** [review]
Good catch !

I think that's a bit of a hack because we match it on all incoming logs. Would it be hard to make it only on docker sources now that we have a decoder instance by tailer ?

**Comment by tt810** [comment]
> Good catch !
> 
> I think that's a bit of a hack because we match it on all incoming logs. Would it be hard to make it only on docker sources now that we have a decoder instance by tailer ?

hmm, decoder is general to all, we will need to do `instanceof` (if there is any) based on the parser passed in.

**Review by achntrl — DISMISSED** [review]

File: pkg/logs/decoder/decoder.go [hunk 3]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
 			maxj = i + d.contentLenLimit
 		} else if inBuf[j] == '\n' {
 			d.lineBuffer.Write(inBuf[i:j])
-			d.sendLine()
-			i = j + 1 // +1 as we skip the `\n`
+			if !matchDockerHeader(d.lineBuffer, d.useDockerParser) {
+				d.sendLine()
+			} else {
+				d.lineBuffer.WriteByte(inBuf[j])
+
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits: empty space

File: pkg/logs/decoder/decoder.go [hunk 4]
```diff
@@ -72,6 +72,14 @@ func InitializeDecoder(source *config.LogSource, parser parser.Parser) *Decoder
 	return New(inputChan, outputChan, lineHandler, lineLimit)
 }
 
+// InitializeDecoderWithDockerParser returns a properly initialized Decoder with docker parser
+// TODO: Refactor decoder to be more specific per source
+func InitializeDecoderWithDockerParser(source *config.LogSource, parser parser.Parser) *Decoder {
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: Why adding the specific logic to the common decoder ? I'd rather duplicate some code and isolate the new logic ; why ? Because if we broke the decoder, it will only be broken for docker logs, plus we enter the `matchDockerHeader` method and do the sanity `useDockerParser` all the time when it's not needed.

File: pkg/logs/decoder/decoder.go [hunk 5]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
 			maxj = i + d.contentLenLimit
 		} else if inBuf[j] == '\n' {
 			d.lineBuffer.Write(inBuf[i:j])
-			d.sendLine()
-			i = j + 1 // +1 as we skip the `\n`
+			if !matchDockerHeader(d.lineBuffer, d.useDockerParser) {
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: I think we should not do anything when it matches, i.e change the condition to:
```
else if inBuf[j] == '\n' && !matchDockerHeader(d.lineBuffer, d.useDockerParser)
```
Unless I am wrong ?

File: pkg/logs/decoder/decoder.go [hunk 6]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+
+			}
+			i = j + 1
 			maxj = i + d.contentLenLimit
 		}
 	}
 	d.lineBuffer.Write(inBuf[i:j])
 }
 
+var dockerHeaderLen = 8
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits: `const`

File: pkg/logs/decoder/decoder.go [hunk 7]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+var dockerHeaderPrefixLen = 4
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits: `const`

File: pkg/logs/decoder/decoder.go [hunk 8]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+var dockerHeaderPrefix1 = []byte{1, 0, 0, 0}
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits: `dockerHeaderStdoutPrefix`

File: pkg/logs/decoder/decoder.go [hunk 9]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+var dockerHeaderPrefix2 = []byte{2, 0, 0, 0}
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits: `dockerHeaderStderrPrefix`

File: pkg/logs/decoder/decoder.go (hunk 4 above)
Comment (ajacquemot) [reply pkg/logs/decoder/decoder.go]: What about moving the specific decoder to the docker package as well ?

File: pkg/logs/decoder/decoder.go [hunk 10]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+// case [1|2 0 0 0 sizeNot10 10 size3 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 10 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 sizeNot10 10]
+// TODO: Refactor decoder to make this check specific to docker source
+func matchDockerHeader(buffer *bytes.Buffer, useDockerParser bool) bool {
+	if !useDockerParser {
+		return false
+	}
+	l := buffer.Len()
+	// len must be less than header length
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: when you add a comment, you should explain why we do things, not what we do. What we do could be explain in the function description.

File: pkg/logs/decoder/decoder.go [hunk 11]
```diff
@@ -119,14 +127,48 @@ func (d *Decoder) decodeIncomingData(inBuf []byte) {
...
+	if l >= dockerHeaderLen || l < dockerHeaderPrefixLen {
+		return false
+	}
+	h := buffer.Bytes()
+
+	return bytes.Equal(h[:dockerHeaderPrefixLen], dockerHeaderPrefix1) ||
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: Please, use `bytes.HasPrefix`

**Review by ajacquemot — CHANGES_REQUESTED** [review]

File: pkg/logs/input/docker/identifier.go [hunk 12]
```diff
@@ -3,8 +3,6 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-2019 Datadog, Inc.
 
-// +build docker
```
Comment (ajacquemot) [inline pkg/logs/input/docker/identifier.go]: this should be kept, we just have to create a specific decoder for docker.

File: pkg/logs/input/docker/parser.go [hunk 13]
```diff
@@ -2,7 +2,6 @@
 // under the Apache License Version 2.0.
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-2019 Datadog, Inc.
-// +build docker
```
Comment (ajacquemot) [inline pkg/logs/input/docker/parser.go]: same

File: pkg/logs/input/docker/parser_test.go [hunk 14]
```diff
@@ -2,7 +2,6 @@
 // under the Apache License Version 2.0.
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-2019 Datadog, Inc.
-// +build docker
```
Comment (ajacquemot) [inline pkg/logs/input/docker/parser_test.go]: same

File: releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml (hunk 2 above)
Comment (ajacquemot) [inline releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml]:
```suggestion
    Empty logs could show up when collecting Docker logs in addition
```

File: releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+    of the actual container logs. This was due to the way the Agent
```
Comment (ajacquemot) [inline releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml]:
```suggestion
    to the actual container logs. This was due to the way the Agent
```

File: releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+---
+fixes:
```
Comment (ajacquemot) [inline releasenotes/notes/logs-fix-empty-log-with-docker-container-64129b051c554ad4.yaml]: I would rephrase this to:
```
Fix a bug where empty logs could be sent when collecting logs from docker containers using the docker socket.
```

**Comment by ajacquemot** [comment]
Finished to review the code, I did not look at the tests yet.
Overall looks good to me, I think you address the problem correctly considering our current design but I have on main concern which is to isolate completely the docker specific decoding logic from the decoder ; I left an idea on how to address this problem.

File: pkg/logs/decoder/decoder.go (hunk 5 above)
Comment (ajacquemot) [reply pkg/logs/decoder/decoder.go]: We could probably have an interface which says when the line is complete so that we limit code duplication

File: pkg/logs/decoder/decoder.go (hunk 4 above)
Comment (tt810) [reply pkg/logs/decoder/decoder.go]: I am not sure replicating the majority of the code for specific source is better. Both way we are hacking this decoder. I would want to keep the code in the same place.

Comment (tt810) [reply pkg/logs/decoder/decoder.go]: The flag `useDockerParser` is assigned when initializing decoder, I don't think it will effect performance to do this check.

File: pkg/logs/decoder/decoder.go (hunk 5 above)
Comment (tt810) [reply pkg/logs/decoder/decoder.go]: Let's try to push it to v6.12.0

File: pkg/logs/input/docker/identifier.go (hunk 12 above)
Comment (tt810) [reply pkg/logs/input/docker/identifier.go]: could you explain why we need to keep this? I understand this is for building, but is it to add more constraint? if we don't put `+build docker`, it won't be included? because the less constraint we have the easier we can test it and linter also can check it(it was ignored by linter).

File: pkg/logs/decoder/decoder.go [hunk 17]
```diff
@@ -48,14 +47,20 @@ func NewOutput(content []byte, status string, rawDataLen int, timestamp string)
 type Decoder struct {
 	InputChan  chan *Input
 	OutputChan chan *Output
-
+	EndLineStrategy
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: I struggled finding a name to isolate the two different logic to send logs either in streaming or batching mode and I came up with the name of the design pattern as well which is I think is a bad pattern.
I think we should probably rename this interface `EndlineMatcher` which provides a `MatchesEndline` method

File: pkg/logs/decoder/strategies.go [hunk 18]
```diff
@@ -0,0 +1,21 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package decoder
```
Comment (ajacquemot) [inline pkg/logs/decoder/strategies.go]: as discussed before, I'd rename this file and class, `EndlineMatcher`

File: pkg/logs/input/docker/decoder.go [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+// Decoder specific to docker container source.
+type Decoder struct {
+	*decoder.Decoder
+}
+
+// InitializeDecoder returns a properly initialized Decoder
+func InitializeDecoder(source *config.LogSource, containerID string) *Decoder {
+
```
Comment (ajacquemot) [inline pkg/logs/input/docker/decoder.go]: nits: in go, we tend to avoid empty space

File: pkg/logs/input/docker/decoder.go [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+package docker
+
+import (
+	"bytes"
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/decoder"
+)
+
+// Decoder specific to docker container source.
+type Decoder struct {
```
Comment (ajacquemot) [inline pkg/logs/input/docker/decoder.go]: here you just need a constructor, not a new decoder.

File: pkg/logs/input/docker/decoder.go [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+// When a newline (in byte is 10) is matching, an additional check need to
+// be done to make sure this is not part of docker header.
+// [1|2 0 0 0 size1 size2 size3 size4], where size1 size2 size3 size4 can be
+// 10 in byte.
+// case [1|2 0 0 0 10 size2 size3 size4]
+// case [1|2 0 0 0 sizeNot10 10 size3 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 10 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 sizeNot10 10]
+func (s *matchHeader) matchDockerHeader(buffer *bytes.Buffer, bs []byte) bool {
```
Comment (ajacquemot) [inline pkg/logs/input/docker/decoder.go]:
```suggestion
func (s *matchHeader) matchesDockerHeader(buffer *bytes.Buffer, bs []byte) bool {
```

File: pkg/logs/input/docker/decoder.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	d.Decoder.Start()
+}
+
+// Stop stops the Decoder
+func (d *Decoder) Stop() {
+	d.Decoder.Stop()
+}
+
+const (
+	dockerHeaderLen       = 8
```
Comment (ajacquemot) [inline pkg/logs/input/docker/decoder.go]: as you're already in the docker package, you can remove all docker prefixes in this file.

File: pkg/logs/input/docker/decoder.go (hunk 21 above)
Comment (ajacquemot) [reply pkg/logs/input/docker/decoder.go]: Why do we need to isolate this method ? I'd find it clearer to have one just method.

File: pkg/logs/input/docker/decoder.go [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+// case [1|2 0 0 0 10 size2 size3 size4]
+// case [1|2 0 0 0 sizeNot10 10 size3 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 10 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 sizeNot10 10]
+func (s *matchHeader) matchDockerHeader(buffer *bytes.Buffer, bs []byte) bool {
+	l := buffer.Len() + len(bs)
+	if l >= dockerHeaderLen || l < dockerHeaderPrefixLen {
+		return false
+	}
+	h := append(buffer.Bytes(), bs...)
```
Comment (ajacquemot) [inline pkg/logs/input/docker/decoder.go]: in most of the logs code base, we have full variable names, can we stick to it please ?

File: pkg/logs/decoder/decoder.go [hunk 24]
```diff
@@ -48,14 +47,20 @@ func NewOutput(content []byte, status string, rawDataLen int, timestamp string)
...
-
+	EndLineMatcher
 	lineBuffer      *bytes.Buffer
 	lineHandler     LineHandler
 	contentLenLimit int
 }
 
 // InitializeDecoder returns a properly initialized Decoder
 func InitializeDecoder(source *config.LogSource, parser parser.Parser) *Decoder {
+	var d = NewDecoderWithStrategy(source, parser, &newLineMatcher{})
```
Comment (ajacquemot) [inline pkg/logs/decoder/decoder.go]: nits:
```
return NewDecoderWithStrategy(source, parser, &newLineMatcher{})
```

**Review by ajacquemot — COMMENTED** [review]
LGTM, thanks for working on this, pretty about the result, one last nits to address.

**Review by ajacquemot — DISMISSED** [review]
LGTM, would be nice to have a quick review from @achntrl as well.

File: pkg/logs/decoder/matcher.go [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package decoder
+
+import "bytes"
+
+// EndLineMatcher defines the criterion to whether to end a line or not.
+type EndLineMatcher interface {
+	Match(buffer *bytes.Buffer, bs []byte, start int, end int) bool
```
Comment (achntrl) [inline pkg/logs/decoder/matcher.go]: Do we really need to leak the buffer here ?

File: pkg/logs/decoder/decoder.go [hunk 26]
```diff
@@ -48,14 +47,19 @@ func NewOutput(content []byte, status string, rawDataLen int, timestamp string)
...
 	contentLenLimit int
 }
 
 // InitializeDecoder returns a properly initialized Decoder
 func InitializeDecoder(source *config.LogSource, parser parser.Parser) *Decoder {
+	return NewDecoderWithEndLineMatcher(source, parser, &newLineMatcher{})
+}
+
+// NewDecoderWithEndLineMatcher initialize a decoder with given endline strategy.
+func NewDecoderWithEndLineMatcher(source *config.LogSource, parser parser.Parser, strategy EndLineMatcher) *Decoder {
```
Comment (achntrl) [inline pkg/logs/decoder/decoder.go]: Nits: `stategy` --> `matcher`

File: pkg/logs/input/docker/decoder.go [hunk 27]
```diff
@@ -0,0 +1,56 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+// +build docker
+
+package docker
```
Comment (achntrl) [inline pkg/logs/input/docker/decoder.go]: Not fan of the packaging of this file, also, I think it should be `matcher.go` instead of `decoder.go`

**Review by achntrl — DISMISSED** [review]

File: pkg/logs/decoder/matcher.go [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package decoder
+
+// EndLineMatcher defines the criterion to whether to end a line or not.
+type EndLineMatcher interface {
+	// Match takes the existing bytes and the bytes to be appended, returns
+	// true if the combination matches the end of line condition.
+	Match(exists []byte, appender []byte, start int, end int) bool
```
Comment (achntrl) [inline pkg/logs/decoder/matcher.go]: I don't mind using bytes.Buffer instead of []byte (I don't know if the `d.lineBuffer.Bytes()` does an allocation of not)
I was wondering if we need this variable at all, or if we can just use `(appender []byte, start int, end int)`, because I think the Docker tailer is going to send full line. But that's a assumption we probably don't want to make

**Review by achntrl — APPROVED** [review]
